float32[] x_coords
float32[] y_coords
float32[] z_coords
int8[] contact_indices